After 304007@main, a network activity is no longer taken on process swap

pvollan opened the pull request:
#### e4ade59d3c143f9f86b82aac74978dd9db21c988
<pre>
After 304007@main, a network activity is no longer taken on process swap
<a href="https://bugs.webkit.org/show_bug.cgi?id=303767">https://bugs.webkit.org/show_bug.cgi?id=303767</a>
<a href="https://rdar.apple.com/166075191">rdar://166075191</a>

Reviewed by Sihui Liu.

This is caused by the fact that NavigationState::didSwapWebProcesses() is called after the process
swap has happened, so we do no longer have information about whether the previous process had a
network activity. This is resolved by checking if a page load is in progress when the process is
being swapped. Additionally, we don&apos;t clear the network activity in WebProcessActivityState::reset(),
since NavigationState has a timer that handles this. This fixes a page load time regression
introduced by 304007@main.

* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::releaseNetworkActivity):
(WebKit::NavigationState::didChangeIsLoading):
(WebKit::NavigationState::didSwapWebProcesses):
* Source/WebKit/UIProcess/WebProcessActivityState.cpp:
(WebKit::WebProcessActivityState::reset):

Canonical link: <a href="https://commits.webkit.org/304180@main">https://commits.webkit.org/304180@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3fc75e45f0504199c0c1d97989c23b3019909bef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134819 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7259 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46071 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142335 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6e4ab03d-2e57-4af7-9aee-d198d4f4790d) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7870 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7104 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/cae04ee1-953c-4a9a-ba9d-7e11528b8ac9) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137766 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5552 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120847 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83854 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/65527500-5d1c-4e94-8f11-2a64919eebfb) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5386 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3002 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2926 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145031 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6932 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39574 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6996 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/5816 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111737 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5228 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117126 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60865 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20808 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6979 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35304 "Passed tests") | | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6768 "Failed to checkout and rebase branch from PR 55048") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70560 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7004 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6877 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

szewai reviewed (commented):
I feel the new flag is redundant (i.e. we are tracking the network activity state in both `WebPageProxy::m_mainFrameProcessActivityState` and `NavigationState::m_hasNetworkActivity`), and it seems error-prone (in `NavigationState::didChangeIsLoading()`, we are setting `NavigationState ::m_hasNetworkActivity` based on state of `WebPageProxy`, but are these two states always sync?)
Can we just invoke `WebPageProxy::takeNetworkActivity()`  after process swap if `WebPageProxy::pageLoadState().isLoading()` is true?

szewai reviewed (approved):
Change LGTM and commit message needs to be updated

pvollan committed:
After 304007@main, a network activity is no longer taken on process swap
https://bugs.webkit.org/show_bug.cgi?id=303767
rdar://166075191

Reviewed by Sihui Liu.

This is caused by the fact that NavigationState::didSwapWebProcesses() is called after the process
swap has happened, so we do no longer have information about whether the previous process had a
network activity. This is resolved by checking if a page load is in progress when the process is
being swapped. Additionally, we don't clear the network activity in WebProcessActivityState::reset(),
since NavigationState has a timer that handles this. This fixes a page load time regression
introduced by 304007@main.

* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::releaseNetworkActivity):
(WebKit::NavigationState::didChangeIsLoading):
(WebKit::NavigationState::didSwapWebProcesses):
* Source/WebKit/UIProcess/WebProcessActivityState.cpp:
(WebKit::WebProcessActivityState::reset):

Canonical link: https://commits.webkit.org/304180@main
